Use server-side filtering for CLI list commands

presztak opened the pull request:
This PR adds support for using server-side filtering for CLI list commands. Closes #10013

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

presztak commented:
I have some doubts:

1. In some cases CLI filtering supports regular expressions e.g.
`lxc list user.blah=h.*`

currently this is not supported on server side. Should we support this on server side?

2. CLI filtering supports case sensitive so for e.g.

`lxc list state=running` and `lxc list state=Running` is equivalent.

This is not supported on server side. Should we support this on server side?

3. CLI filtering supports abbreviated to the smallest unambiguous identifier for configuration items e.g.

`lxc list u.blah=abc`

Should we support this on server side or let cli to take care of it?

4. CLI allows us to filter by `ExpandedConfig` in case of instances and `Properties` in case of images without specify ExpandedConfig/Properties field. So we can do:

`lxc list u.blah=abc`

but to be able to do this on server side we need to specify ExpandedConfig/Properties field. So we need to do:

`lxc list expanded_config.u.blah=abc`

What will be the best way to support this and have backward compatibility?

lxc-jenkins commented:
Documentation preview available at: https://linuxcontainers.org/lxd/docs/pr.10140/

presztak commented:
@tomponline @stgraber any thoughts regarding my above comments?

tomponline commented:
@presztak sorry for not looking at this right now we've been very busy with the 5.0 LTS release and follow up fixes. Will look ASAP, thanks! :)

stgraber commented on client/lxd_images.go:
```diff
@@ -930,3 +951,28 @@ func (r *ProtocolLXD) ExportImage(fingerprint string, image api.ImageExportPost)
 }
+
+// GetServerSupportedFilters returns two lists: one with filters supported by server and second one with not supported.
+func (r *ProtocolLXD) GetServerSupportedFilters(filters []string, i interface{}) ([]string, []string) {
```
This feels a bit odd. I think it'd probably be simpler to turn this into an internal `validateFilters(filters []string) error` which then errors out if missing the extension or if the filter syntax is invalid.

Ultimately it's going to be up to the caller of those functions to pass in valid filters and to the server to return an error when something unsupported is passed.

presztak replied:
I've moved this method to `lxc/utils`.

stgraber commented on client/lxd_images.go:
```diff
@@ -33,6 +34,26 @@ func (r *ProtocolLXD) GetImages() ([]api.Image, error) {
 }
 
+// GetFilteredImages returns a filtered list of available images as Image structs
+func (r *ProtocolLXD) GetFilteredImages(filters []string) ([]api.Image, error) {
```
Maybe `GetImagesWithFilter` instead?

presztak replied:
Fixed.

stgraber commented on client/lxd_instances.go:
```diff
@@ -116,6 +116,31 @@ func (r *ProtocolLXD) GetInstances(instanceType api.InstanceType) ([]api.Instanc
 }
 
+// GetFilteredInstances returns a filtered list of instances.
+func (r *ProtocolLXD) GetFilteredInstances(instanceType api.InstanceType, filters []string) ([]api.Instance, error) {
```
Maybe `GetInstancesWithFilter` instead?

presztak replied:
Fixed.

stgraber commented on client/lxd_instances.go:
```diff
@@ -141,6 +166,36 @@ func (r *ProtocolLXD) GetInstancesAllProjects(instanceType api.InstanceType) ([]
 }
 
+// GetFilteredInstancesAllProjects returns a filtered list of instances from all projects.
+func (r *ProtocolLXD) GetFilteredInstancesAllProjects(instanceType api.InstanceType, filters []string) ([]api.Instance, error) {
```
Maybe `GetInstancesAllProjectsWithFilter` instead?

presztak replied:
Fixed.

stgraber commented on client/lxd_instances.go:
```diff
@@ -181,6 +236,35 @@ func (r *ProtocolLXD) GetInstancesFull(instanceType api.InstanceType) ([]api.Ins
 }
 
+// GetFilteredInstancesFull returns a filtered list of instances including snapshots, backups and state.
+func (r *ProtocolLXD) GetFilteredInstancesFull(instanceType api.InstanceType, filters []string) ([]api.InstanceFull, error) {
```
Maybe `GetInstancesFullWithFilter` instead?

presztak replied:
Fixed.

stgraber commented on client/lxd_instances.go:
```diff
@@ -210,6 +294,40 @@ func (r *ProtocolLXD) GetInstancesFullAllProjects(instanceType api.InstanceType)
 }
 
+// GetFilteredInstancesFullAllProjects returns a filtered list of instances including snapshots, backups and state from all projects.
+func (r *ProtocolLXD) GetFilteredInstancesFullAllProjects(instanceType api.InstanceType, filters []string) ([]api.InstanceFull, error) {
```
Maybe `GetInstancesFullAllProjectsWithFilter` instead (this is getting long ;))?

stgraber commented on client/lxd_storage_volumes.go:
```diff
@@ -50,6 +50,26 @@ func (r *ProtocolLXD) GetStoragePoolVolumes(pool string) ([]api.StorageVolume, e
 }
 
+// GetFilteredStoragePoolVolumes returns a filtered list of StorageVolume entries for the provided pool
+func (r *ProtocolLXD) GetFilteredStoragePoolVolumes(pool string, filters []string) ([]api.StorageVolume, error) {
```
Maybe `GetStoragePoolVolumesWithFilter` instead?

presztak replied:
Fixed.

stgraber commented on client/simplestreams_images.go:
```diff
@@ -21,6 +21,11 @@ func (r *ProtocolSimpleStreams) GetImages() ([]api.Image, error) {
 }
 
+// GetFilteredImages returns a filtered list of available images as Image structs
+func (r *ProtocolSimpleStreams) GetFilteredImages(filters []string) ([]api.Image, error) {
```
I think this should return an error instead.

stgraber commented on client/simplestreams_images.go:
```diff
@@ -282,3 +287,8 @@ func (r *ProtocolSimpleStreams) GetImageAliasArchitectures(imageType string, nam
 	return nil, fmt.Errorf("Exporting images is not supported by the simplestreams protocol")
 }
+
+// GetServerSupportedFilters returns two lists: one with filters supported by server and second one with not supported.
+func (r *ProtocolSimpleStreams) GetServerSupportedFilters(filters []string, i interface{}) ([]string, []string) {
+	return []string{}, filters
+}
```
This should go away if we move to `validateFilters` or something like that instead.

presztak replied:
I've moved this method to `lxc/utils`.

stgraber commented on lxc/list.go:
```diff
@@ -512,41 +512,45 @@ func (c *cmdList) Run(cmd *cobra.Command, args []string) error {
 	// Get the list of instances
 	var cts []api.Instance
 	var ctslist []api.Instance
+	serverFilters, clientFilters := d.GetServerSupportedFilters(filters, api.Instance{})
```
Ah, I see why that function exists now :)

I'm still not sure it really fits with the `client` package though, so maybe easier to keep it in lxc/utils for now?

presztak commented:
Hey @stgraber,

I've fixed all yours comments. Just one thing left, I didn't rename `GetServerSupportedFilters` to `validateFilters` but I've moved it to `lxc/utils`.
To could proceed this issue I still need answers for my above questions :)

stgraber commented:
> 1. In some cases CLI filtering supports regular expressions e.g.
>        `lxc list user.blah=h.*`
>        currently this is not supported on server side. Should we support this on server side?

Seems fine to support server-side.



> 2\. CLI filtering supports case sensitive so for e.g.
>         `lxc list state=running` and `lxc list state=Running` is equivalent.
>         This is not supported on server side. Should we support this on server side?

Happy to have state be case insensitive on the server side.



> 3\. CLI filtering supports abbreviated to the smallest unambiguous identifier for configuration items e.g.
>         `lxc list u.blah=abc`
>         Should we support this on server side or let cli to take care of it?

I think that's best done by the client.



> 4\. CLI allows us to filter by `ExpandedConfig` in case of instances and `Properties` in case of images without specify ExpandedConfig/Properties field. So we can do:
>         `lxc list u.blah=abc`
>         but to be able to do this on server side we need to specify ExpandedConfig/Properties field. So we need to do:
>         `lxc list expanded_config.u.blah=abc`
>         What will be the best way to support this and have backward compatibility?

Can we do the naive thing of assuming that anything with a `.` in it is going to be a filter for ExpandedConfig?
Properties should be flat so no dots in them.

tomponline commented:
@presztak please can you rebase, and then it will be ready to merge if @stgraber is happy with it.

stgraber commented:
Did the rebase.

stgraber commented:
jenkins: test this please

stgraber commented:
Thanks!
